File: backend/app/analysis/indicators.py
"""纯 Python 技术指标(无 pandas/numpy 依赖)。

输入均为收盘价/价格序列 list[float],返回与输入等长的序列(前导不足处为 None),
或标量。供 signals.py 与 API 使用;前端 TS 版本逻辑一致(frontend/lib/indicators.ts)。
"""
from __future__ import annotations

from typing import Optional


def sma(values: list[float], period: int) -> list[Optional[float]]:
    out: list[Optional[float]] = [None] * len(values)
    if period <= 0:
        return out
    run = 0.0
    for i, v in enumerate(values):
        run += v
        if i >= period:
            run -= values[i - period]
        if i >= period - 1:
            out[i] = run / period
    return out


def ema(values: list[float], period: int) -> list[Optional[float]]:
    out: list[Optional[float]] = [None] * len(values)
    if period <= 0 or len(values) < period:
        return out
    k = 2.0 / (period + 1)
    # 用前 period 个的 SMA 作为种子
    seed = sum(values[:period]) / period
    out[period - 1] = seed
    prev = seed
    for i in range(period, len(values)):
        prev = values[i] * k + prev * (1 - k)
        out[i] = prev
    return out


def rsi(values: list[float], period: int = 14) -> list[Optional[float]]:
    out: list[Optional[float]] = [None] * len(values)
    if len(values) <= period:
        return out
    gains = 0.0
    losses = 0.0
    for i in range(1, period + 1):
        d = values[i] - values[i - 1]
        gains += max(d, 0.0)
        losses += max(-d, 0.0)
    avg_gain = gains / period
    avg_loss = losses / period
    out[period] = 100.0 if avg_loss == 0 else 100.0 - 100.0 / (1 + avg_gain / avg_loss)
    for i in range(period + 1, len(values)):
        d = values[i] - values[i - 1]
        avg_gain = (avg_gain * (period - 1) + max(d, 0.0)) / period
        avg_loss = (avg_loss * (period - 1) + max(-d, 0.0)) / period
        out[i] = 100.0 if avg_loss == 0 else 100.0 - 100.0 / (1 + avg_gain / avg_loss)
    return out


def macd(values: list[float], fast: int = 12, slow: int = 26, signal: int = 9):
    """返回 (macd_line, signal_line, histogram)。"""
    ef = ema(values, fast)
    es = ema(values, slow)
    line: list[Optional[float]] = [
        (a - b) if (a is not None and b is not None) else None for a, b in zip(ef, es)
    ]
    # 对 line 中非 None 段做 EMA(signal)
    idx = [i for i, v in enumerate(line) if v is not None]
    sig: list[Optional[float]] = [None] * len(values)
    if len(idx) >= signal:
        seq = [line[i] for i in idx]  # type: ignore[misc]
        es2 = ema(seq, signal)  # type: ignore[arg-type]
        for j, i in enumerate(idx):
            sig[i] = es2[j]
    hist: list[Optional[float]] = [
        (l - s) if (l is not None and s is not None) else None for l, s in zip(line, sig)
    ]
    return line, sig, hist


def bollinger(values: list[float], period: int = 20, mult: float = 2.0):
    """返回 (mid, upper, lower)。"""
    mid = sma(values, period)
    upper: list[Optional[float]] = [None] * len(values)
    lower: list[Optional[float]] = [None] * len(values)
    for i in range(len(values)):
        if i >= period - 1:
            window = values[i - period + 1 : i + 1]
            m = mid[i]
            if m is None:
                continue
            var = sum((x - m) ** 2 for x in window) / period
            sd = var ** 0.5
            upper[i] = m + mult * sd
            lower[i] = m - mult * sd
    return mid, upper, lower


def atr(highs: list[float], lows: list[float], closes: list[float], period: int = 14) -> Optional[float]:
    n = len(closes)
    if n <= period:
        return None
    trs: list[float] = []
    for i in range(1, n):
        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        trs.append(tr)
    # 简单平均(Wilder 平滑亦可,这里取近 period TR 均值)
    return sum(trs[-period:]) / period


def pct_return(values: list[float], lookback: int) -> Optional[float]:
    if len(values) <= lookback or values[-1 - lookback] == 0:
        return None
    return (values[-1] / values[-1 - lookback] - 1.0) * 100.0


def volatility(values: list[float], lookback: int = 20) -> Optional[float]:
    """近 lookback 日对数收益的年化波动率(%)。"""
    import math

    if len(values) <= lookback:
        return None
    rets = []
    for i in range(len(values) - lookback, len(values)):
        if values[i - 1] > 0:
            rets.append(math.log(values[i] / values[i - 1]))
    if len(rets) < 2:
        return None
    mean = sum(rets) / len(rets)
    var = sum((r - mean) ** 2 for r in rets) / (len(rets) - 1)
    return (var ** 0.5) * (252 ** 0.5) * 100.0


def last(values: list[Optional[float]]) -> Optional[float]:
    for v in reversed(values):
        if v is not None:
            return v
    return None

File: frontend/lib/indicators.ts
// 纯 TS 技术指标(与 backend/app/analysis/indicators.py 逻辑一致)。
export type Maybe = number | null;

export function sma(v: number[], p: number): Maybe[] {
  const out: Maybe[] = Array(v.length).fill(null);
  if (p <= 0) return out;
  let run = 0;
  for (let i = 0; i < v.length; i++) {
    run += v[i];
    if (i >= p) run -= v[i - p];
    if (i >= p - 1) out[i] = run / p;
  }
  return out;
}

export function ema(v: number[], p: number): Maybe[] {
  const out: Maybe[] = Array(v.length).fill(null);
  if (p <= 0 || v.length < p) return out;
  const k = 2 / (p + 1);
  let prev = v.slice(0, p).reduce((a, b) => a + b, 0) / p;
  out[p - 1] = prev;
  for (let i = p; i < v.length; i++) {
    prev = v[i] * k + prev * (1 - k);
    out[i] = prev;
  }
  return out;
}

export function rsi(v: number[], p = 14): Maybe[] {
  const out: Maybe[] = Array(v.length).fill(null);
  if (v.length <= p) return out;
  let g = 0, l = 0;
  for (let i = 1; i <= p; i++) {
    const d = v[i] - v[i - 1];
    g += Math.max(d, 0);
    l += Math.max(-d, 0);
  }
  let ag = g / p, al = l / p;
  out[p] = al === 0 ? 100 : 100 - 100 / (1 + ag / al);
  for (let i = p + 1; i < v.length; i++) {
    const d = v[i] - v[i - 1];
    ag = (ag * (p - 1) + Math.max(d, 0)) / p;
    al = (al * (p - 1) + Math.max(-d, 0)) / p;
    out[i] = al === 0 ? 100 : 100 - 100 / (1 + ag / al);
  }
  return out;
}

export function macd(v: number[], fast = 12, slow = 26, signal = 9) {
  const ef = ema(v, fast), es = ema(v, slow);
  const line: Maybe[] = ef.map((a, i) => (a !== null && es[i] !== null ? a - (es[i] as number) : null));
  const idx: number[] = [];
  line.forEach((x, i) => x !== null && idx.push(i));
  const sig: Maybe[] = Array(v.length).fill(null);
  if (idx.length >= signal) {
    const seq = idx.map((i) => line[i] as number);
    const es2 = ema(seq, signal);
    idx.forEach((i, j) => (sig[i] = es2[j]));
  }
  const hist: Maybe[] = line.map((x, i) => (x !== null && sig[i] !== null ? x - (sig[i] as number) : null));
  return { line, sig, hist };
}

export function bollinger(v: number[], p = 20, mult = 2) {
  const mid = sma(v, p);
  const up: Maybe[] = Array(v.length).fill(null);
  const lo: Maybe[] = Array(v.length).fill(null);
  for (let i = p - 1; i < v.length; i++) {
    const m = mid[i];
    if (m === null) continue;
    const w = v.slice(i - p + 1, i + 1);
    const sd = Math.sqrt(w.reduce((a, x) => a + (x - m) ** 2, 0) / p);
    up[i] = m + mult * sd;
    lo[i] = m - mult * sd;
  }
  return { mid, up, lo };
}

export function last(a: Maybe[]): Maybe {
  for (let i = a.length - 1; i >= 0; i--) if (a[i] !== null) return a[i];
  return null;
}

export function pctReturn(v: number[], lb: number): Maybe {
  if (v.length <= lb || v[v.length - 1 - lb] === 0) return null;
  return (v[v.length - 1] / v[v.length - 1 - lb] - 1) * 100;
}

/** ATR(p):平均真实波幅(Wilder 平滑)。bars 需含 high/low/close。 */
export function atr(highs: number[], lows: number[], closes: number[], p = 14): Maybe {
  const n = closes.length;
  if (n <= p) return null;
  const trs: number[] = [];
  for (let i = 1; i < n; i++) {
    trs.push(Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1]),
    ));
  }
  let a = trs.slice(0, p).reduce((s, x) => s + x, 0) / p;
  for (let i = p; i < trs.length; i++) a = (a * (p - 1) + trs[i]) / p;
  return a;
}

/** KDJ(9,3,3):返回最新 K/D/J。 */
export function kdj(highs: number[], lows: number[], closes: number[], p = 9): { k: Maybe; d: Maybe; j: Maybe } {
  const n = closes.length;
  if (n < p) return { k: null, d: null, j: null };
  let k = 50, d = 50;
  for (let i = p - 1; i < n; i++) {
    const hh = Math.max(...highs.slice(i - p + 1, i + 1));
    const ll = Math.min(...lows.slice(i - p + 1, i + 1));
    const rsv = hh === ll ? 50 : ((closes[i] - ll) / (hh - ll)) * 100;
    k = (2 / 3) * k + (1 / 3) * rsv;
    d = (2 / 3) * d + (1 / 3) * k;
  }
  return { k, d, j: 3 * k - 2 * d };
}

/** 量比:最新一根成交量 / 近 p 根均量(>1 放量,<1 缩量)。 */
export function volumeRatio(volumes: number[], p = 20): Maybe {
  if (volumes.length < p + 1) return null;
  const recent = volumes[volumes.length - 1];
  const base = volumes.slice(-p - 1, -1).reduce((s, x) => s + x, 0) / p;
  return base > 0 ? recent / base : null;
}
